src/Browscap/Coverage/Processor.php

```diff
@@ -0,0 +1,571 @@
+    {
+        $this->resourceDir = $resourceDir;
+
+        $this->symbolLookup = array_flip($this->symbols);
+    }
+
+    /**
+     * Process the directory of JSON files using the collected pattern ids
+     *
+     * @param array $coveredIds
+     */
+    public function process(array $coveredIds)
+    {
+        $this->setCoveredPatternIds($coveredIds);
+
+        $iterator = new \RecursiveDirectoryIterator($this->resourceDir);
+
+        foreach (new \RecursiveIteratorIterator($iterator) as $file) {
+            /** @var $file \SplFileInfo */
+            if (!$file->isFile() || $file->getExtension() !== 'json') {
+                continue;
+            }
+
+            $patternFileName = substr($file->getPathname(), strpos($file->getPathname(), 'resources/'));
+
+            if (!isset($this->coveredIds[$patternFileName])) {
+                $this->coveredIds[$patternFileName] = [];
+            }
+
+            $this->coverage[$patternFileName] = $this->processFile(
+                $patternFileName,
+                file_get_contents($file->getPathname()),
+                $this->coveredIds[$patternFileName]
+            );
+        }
+    }
+
+    /**
+     * Write the coverage data in JSON format to specified filename
+     *
+     * @param string $fileName
+     */
+    public function write($fileName)
+    {
+        file_put_contents(
+            $fileName,
+            // str_replace here is to convert empty arrays into empty JS objects, which is expected by
+            // codecov.io. Owner of the service said he was going to patch it, haven't tested since
+            // Note: Can't use JSON_FORCE_OBJECT here as we **do** want arrays for the 'b' structure
+            // which FORECE_OBJECT turns into objects, breaking at least the Istanbul coverage reporter
+            str_replace('[]', '{}', json_encode($this->coverage, JSON_UNESCAPED_SLASHES))
+        );
+    }
+
+    /**
+     * Stores passed in pattern ids, grouping them by file first
+     *
+     * @param array $coveredIds
+     */
+    public function setCoveredPatternIds(array $coveredIds)
+    {
+        $this->coveredIds = $this->groupIdsByFile($coveredIds);
+    }
+
+    /**
+     * Returns the grouped pattern ids previously set
+     *
+     * @return array
+     */
+    public function getCoveredPatternIds()
+    {
+        return $this->coveredIds;
+    }
+
+    /**
+     * Process an individual file for coverage data using covered ids
+     *
+     * @param  string $file
+     * @param  string $contents
+     * @param  array  $coveredIds
+     * @return array
+     */
+    public function processFile($file, $contents, array $coveredIds)
+    {
+        $coverage = [
+            'path' => $file,
+            'statementMap' => [],
+            'fnMap' => [],
+            'branchMap' => [],
+            's' => [],
+            'b' => [],
+            'f' => [],
+        ];
+
+        $lines = explode("\n", $contents);
+
+        $u = null;
```

Comment:
I used $u, $p, $d and $c to match the letter prefix used in the PatternId, since they're used to represent the data of that spot in the id.  Still like me to change?

Most likely I would change to something like this:

$u -> $userAgentPosition
$c -> $childNumber
$d -> $deviceMatch
$p -> $platformName